037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+Such a rebalancing can happen right after the scale up to move some replicas to the newly added brokers or right before the scale down to move replicas off of the brokers to be removed.
+
+## Current situation
+
+Currently, there is no automatic rebalancing when the user scales the cluster up or down.
+It means that:
+
+* after scaling up the cluster, in order to rebalance the cluster and move some replicas to the newly added brokers, the user has to create a `KafkaRebalance` custom resource using the `spec.mode: add-brokers` mode and add the new broker IDs to the the `spec.brokers` field.
+* before scaling down the cluster, in order to rebalance the cluster and move replicas off of the brokers to be removed, the user has to create a `KafkaRebalance` custom resource using the `spec.mode: remove-brokers` mode and add the broker IDs of the brokers to be removed to the `spec.brokers` field.
+
+For more details, the proposal [035](https://github.com/strimzi/proposals/blob/main/035-rebalance-types-scaling-brokers.md) introduced the rebalancing modes mentioned above.
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two step process.
+
+* scaling up the cluster, then running a rebalance.
+* running a rebalance, then scaling down the cluster.
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster by changing the `spec.kafka.replicas` field and then having the operator taking care of running a rebalancing operation automatically if requested.
+
+Taking into account the `StrimziPodSets` usage (instead of the `StatefulSets`), which will be the default in the future, a more general use case will be about adding new brokers and/or removing existing ones, not talking just about scaling up and down anymore.
+
+This proposal takes into account this future plan by referring to more general operations like add and remove brokers instead of scaling up and down.
+
+## Proposal
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the configuration for the auto-rebalancing but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance` one.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, of adding the configuration to the `Kafka` custom resource, this will avoid polluting the `Kafka` custom resource and increasing its size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+  labels:
+    strimzi.io/cluster: my-cluster
```

Comment:
It would be ignored as it happens for a `KafkaRebalance` custom resource and logging a warning.
In order to provide more info to the user, a warning condition could be added to the `Kafka` status. Wdyt?

Reply:
KafkaRebalance is acted upon. But this is just a template. From my point of view:
* You make this more complicated then it has to be (you need to add the logic to check the label, have error handling for it etc.)
* You do not allow to share the the template between different clusters

Unless we allow to reference them across namespaces, the sharing of the templates is probably not major use-case. But I wonder more what do you get in exchange for the first point. It seems to me that without the label on the template resource you:
* Have less code complexity and make your life easier
* Do not loose any actual fucntionality or added value
* Possibly in some edge cases enable a new feature (sharing the template between clusters)

Reply:
Yeah maybe I was adding useless complications on this. Let's remove the cluster label (at least for now).